lib/spells/BattleSpellMechanics.cpp

```diff
@@ -425,16 +432,50 @@ void BattleSpellMechanics::beforeCast(BattleSpellCast & sc, vstd::RNG & rng, con
 		});
 	}
 
-	if(mode == Mode::MAGIC_MIRROR)
+	for(const auto * unit : resisted)
+		sc.resistedCres.insert(unit->unitId());
+}
+
+bool BattleSpellMechanics::isReflected(const battle::Unit * unit, vstd::RNG & rng)
+{
+	auto range = owner -> getLevelInfo(getRangeLevel()).range;
+	const auto directSpellRange = boost::assign::list_of(0);
+	const std::string magicMirrorCacheStr = "type_MAGIC_MIRROR";
+	static const auto magicMirrorSelector = Selector::type()(BonusType::MAGIC_MIRROR);
+
+	bool spellIsDirect = !isMassive() && owner -> getLevelInfo(getRangeLevel()).range == directSpellRange;
+	bool spellIsReflectable = spellIsDirect && (mode == Mode::HERO || mode == Mode::MAGIC_MIRROR) && isNegativeSpell();
+	bool targetCanReflectSpell = spellIsReflectable && unit->getAllBonuses(Selector::type()(BonusType::MAGIC_MIRROR))->size()>0;
+	return targetCanReflectSpell && rng.nextInt(0, 99) < unit->valOfBonuses(magicMirrorSelector, magicMirrorCacheStr);
+}
+
+void BattleSpellMechanics::reflect(BattleSpellCast & sc, vstd::RNG & rng, const battle::Unit* unit)
+{
+	auto otherSide = battle()->otherSide(unit->unitSide());
+	auto newTarget = getRandomUnit(rng, otherSide);
+	auto reflectedTo = newTarget->getPosition();
+
+	mode = Mode::MAGIC_MIRROR;
+	sc.reflectedCres.insert(unit->unitId());
+	sc.tile = reflectedTo;
+
+	if (!isReceptive(newTarget))
 	{
-		if(caster->getHeroCaster() == nullptr)
-		{
-			sc.reflectedCres.insert(caster->getCasterUnitId());
-		}
+		sc.resistedCres.insert(newTarget->unitId());    //A spell can be reflected to then resisted by an immune unit. Consistent with the original game.
 	}
 
-	for(const auto * unit : resisted)
-		sc.resistedCres.insert(unit->unitId());
+	beforeCast(sc, rng,
+		boost::assign::list_of(reflectedTo));
+}
+
+const battle::Unit* BattleSpellMechanics::getRandomUnit(vstd::RNG & rng, BattleSide & side)
+{
+	auto targets = battle()->getBattle()->getUnitsIf([this, & side](const battle::Unit * unit)
```

Comment:
FYI: We have various automatic checks for PR's, mostly compiling on various platforms / various compilers and one of those has failed here - due to a warning that only present on clang:
https://github.com/vcmi/vcmi/actions/runs/16124607314/job/45498525834?pr=5897#step:16:1773
```suggestion
	auto targets = battle()->getBattle()->getUnitsIf([&side](const battle::Unit * unit)
```